client/state/signup/progress/reducer.js

```diff
@@ -0,0 +1,123 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { get, find, map } from 'lodash';
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import stepsConfig from 'signup/config/steps-pure';
+import flows from 'signup/config/flows-pure';
+import {
+	SIGNUP_COMPLETE_RESET,
+	SIGNUP_PROGRESS_COMPLETE_STEP,
+	SIGNUP_PROGRESS_INVALIDATE_STEP,
+	SIGNUP_PROGRESS_PROCESS_STEP,
+	SIGNUP_PROGRESS_REMOVE_UNNEEDED_STEPS,
+	SIGNUP_PROGRESS_SAVE_STEP,
+	SIGNUP_PROGRESS_SET,
+	SIGNUP_PROGRESS_SUBMIT_STEP,
+} from 'state/action-types';
+import { createReducer } from 'state/utils';
+import { schema } from './schema';
+
+const debug = debugFactory( 'calypso:state:signup:progress:reducer' );
+
+//
+// Action handlers
+//
+function overwriteSteps( state, { steps = [] } ) {
+	// When called without action.steps, this is basically a state reset function.
+	return Array.isArray( steps ) ? steps : [];
+}
+
+function completeStep( state, { step } ) {
+	return updateStep( state, { ...step, status: 'completed' } );
+}
+
+function invalidateStep( state, { step, errors } ) {
+	const newStep = { ...step, errors, status: 'invalid' };
+	if ( find( state, { stepName: newStep.stepName } ) ) {
+		return updateStep( state, newStep );
+	}
+	return addStep( state, newStep );
+}
+
+export function processStep( state, { step } ) {
+	return updateStep( state, { ...step, status: 'processing' } );
+}
+
+function removeUnneededSteps( state, { flowName } ) {
+	const flowSteps = get( flows, `${ flowName }.steps`, [] );
+	return state.filter( step => flowSteps.includes( step.stepName ) );
+}
+
+function saveStep( state, { step } ) {
+	if ( find( state, { stepName: step.stepName } ) ) {
+		return updateStep( state, step );
+	}
+
+	return addStep( state, { ...step, status: 'in-progress' } );
+}
+
+function submitStep( state, { step } ) {
+	const stepHasApiRequestFunction = get( stepsConfig, `${ step.stepName }.apiRequestFunction` );
+	const status = stepHasApiRequestFunction ? 'pending' : 'completed';
+
+	if ( find( state, { stepName: step.stepName } ) ) {
+		return updateStep( state, { ...step, status } );
+	}
+
+	return addStep( state, { ...step, status } );
+}
+
+//
+// Helper Functions
+//
+export function addStep( state, step ) {
```

Comment:
This export (and also `updateStep` and `processStep`) exists only for tests, doesn't it? What if the module exported only one thing, the reducer? That's the only export that's really needed by the app. And it's a unit that's small enough to be unit tested as a whole. `addStep` already gets a complete test coverage with the actions that use it. There's no need to test it separately.

I find it hard to figure out modules that export almost everything. When reading the code, I don't know what's the public API of the module.

Besides this little nit, the code looks pretty solid.

What is your plan to handle Redux state persistence for signup? There are two special behaviors that are tricky to implement with our Redux persistence helpers:
- the persisted progress store should be ignored if the current flow name doesn't match the persisted one (see calls to `fetchCachedSignup`)
- the signup flow is the only Calypso feature that crosses the logged-out/logged-in boundary. The store persisted in logged-out state won't be loaded in the logged-in state.

Reply:
> This export (and also `updateStep` and `processStep`) exists only for tests, doesn't it? (@jsnajdr)

Correct. However, `processStep` actually shouldn't be exported; I'll fix that.

> What if the module exported only one thing, the reducer? That's the only export that's really needed by the app. And it's a unit that's small enough to be unit tested as a whole. addStep already gets a complete test coverage with the actions that use it. There's no need to test it separately. (@jsnajdr)

Good point. I originally architected the reducer with the assumption that most action handlers would essentially be a call to `updateStep`. However, that ended up not being the case.

I'll look into eliminating excess exports.

> What is your plan to handle Redux state persistence for signup? There are two special behaviors that are tricky to implement with our Redux persistence helpers:  (@jsnajdr)

> (1) the signup flow is the only Calypso feature that crosses the logged-out/logged-in boundary. The store persisted in logged-out state won't be loaded in the logged-in state.

I'm working off the assumption that the progress store (and the dependency store) is only useful to the logged-out user while they are progressing through the signup flow. Once they complete the flow by creating an account, discarding the logged-out state sounds like the right way to go. 

> (2) There are two special behaviors that are tricky to implement with our Redux persistence helpers: the persisted progress store should be ignored if the current flow name doesn't match the persisted one (see calls to `fetchCachedSignup`) (@jsnajdr)

@taggon recently introduced [a function](https://github.com/Automattic/wp-calypso/commit/aa21a468d1c17aa6267448c8368fcc815cc784b7) to discard steps not associated with the current flow. 

I intend to invoke this function in the constructor of the [`signup flow controller's constructor (where fetchCachedSignup is being invoked)`](https://github.com/Automattic/wp-calypso/blob/4df0c252c0e43aa98749ac5fdd20ff7503198db0/client/lib/signup/flow-controller.js#L63) to clear out any progress not associated with the current flow. Alternatively, we can indiscriminately clear out the progress state when initializing a new signup flow controller.

Thoughts?